File: mm-php/src/io/metamessage/mm/WireEncoder.php
<?php

namespace io\metamessage\mm;

class WireEncoder {
    private GrowableByteBuf $buf;

    public function __construct() {
        $this->buf = new GrowableByteBuf();
    }

    public function toByteArray(): array {
        return $this->buf->copyRange(0, $this->buf->length());
    }

    public function finishTakeLast(int $writtenFromEnd): array {
        $end = $this->buf->length();
        return $this->buf->copyRange($end - $writtenFromEnd, $end);
    }

    public function reset(): void {
        $this->buf->reset();
    }

    public function size(): int {
        return $this->buf->length();
    }

    public function encodeSimple(int $simpleValue): int {
        $start = $this->buf->length();
        $this->buf->write(Prefix::SIMPLE | $simpleValue);
        return $this->buf->length() - $start;
    }

    public function encodeBool(bool $v): int {
        return $this->encodeSimple($v ? SimpleValue::TRUE : SimpleValue::FALSE);
    }

    public function encodeInt64(int $v): int {
        if ($v >= 0) {
            return $this->encodeUintWithPrefix(Prefix::POSITIVE_INT, $v);
        } else {
            $uv = $v === PHP_INT_MIN ? PHP_INT_MAX + 1 : -$v;
            return $this->encodeUintWithPrefix(Prefix::NEGATIVE_INT, $uv);
        }
    }

    public function encodeUint64(int $uv): int {
        if ($uv < 0) {
            throw new \Exception('Expected unsigned');
        }
        return $this->encodeUintWithPrefix(Prefix::POSITIVE_INT, $uv);
    }

    private function encodeUintWithPrefix(int $prefix, int $uv): int {
        $start = $this->buf->length();
        if ($uv < WireConstants::INT_LEN_1) {
            $this->buf->write($prefix | $uv);
        } elseif ($uv <= WireConstants::MAX_1) {
            $this->buf->write($prefix | WireConstants::INT_LEN_1, $uv);
        } elseif ($uv <= WireConstants::MAX_2) {
            $this->buf->write($prefix | WireConstants::INT_LEN_2, ($uv >> 8) & 0xFF, $uv & 0xFF);
        } elseif ($uv <= WireConstants::MAX_3) {
            $this->buf->write($prefix | WireConstants::INT_LEN_3, ($uv >> 16) & 0xFF, ($uv >> 8) & 0xFF, $uv & 0xFF);
        } elseif ($uv <= WireConstants::MAX_4) {
            $this->buf->write($prefix | WireConstants::INT_LEN_4, ($uv >> 24) & 0xFF, ($uv >> 16) & 0xFF, ($uv >> 8) & 0xFF, $uv & 0xFF);
        } else {
            throw new \Exception('Integer too large');
        }
        return $this->buf->length() - $start;
    }

    public function encodeFloatString(string $s): int {
        $p = FloatCodec::parseDecimalString($s);
        $start = $this->buf->length();
        $sign = Prefix::FLOAT;
        if ($p['negative']) {
            $sign |= WireConstants::FLOAT_NEG_MASK;
        }
        if ($p['exponent'] === -1 && $p['mantissa'] <= 7) {
            $this->buf->write($sign | $p['mantissa']);
        } else {
            $mantissa = $p['mantissa'];
            if ($mantissa <= WireConstants::MAX_1) {
                $this->buf->write($sign | WireConstants::FLOAT_LEN_1, $p['exponent'], $mantissa);
            } elseif ($mantissa <= WireConstants::MAX_2) {
                $this->buf->write($sign | WireConstants::FLOAT_LEN_2, $p['exponent'], ($mantissa >> 8) & 0xFF, $mantissa & 0xFF);
            } else {
                throw new \Exception('Float mantissa too large');
            }
        }
        return $this->buf->length() - $start;
    }

    public function encodeString(string $s): int {
        $utf = unpack('C*', $s);
        $length = count($utf);
        $start = $this->buf->length();
        $sign = Prefix::STRING;
        if ($length < WireConstants::STRING_LEN_1) {
            $this->buf->write($sign | $length, ...$utf);
        } elseif ($length < WireConstants::MAX_1) {
            $this->buf->write($sign | WireConstants::STRING_LEN_1, $length, ...$utf);
        } elseif ($length < WireConstants::MAX_2) {
            $this->buf->write($sign | WireConstants::STRING_LEN_2, ($length >> 8) & 0xFF, $length & 0xFF, ...$utf);
        } else {
            throw new \Exception('String too long');
        }
        return $this->buf->length() - $start;
    }

    public function encodeBytes(array $data): int {
        $length = count($data);
        $start = $this->buf->length();
        $sign = Prefix::BYTES;
        if ($length < WireConstants::BYTES_LEN_1) {
            $this->buf->write($sign | $length, ...$data);
        } elseif ($length < WireConstants::MAX_1) {
            $this->buf->write($sign | WireConstants::BYTES_LEN_1, $length, ...$data);
        } elseif ($length < WireConstants::MAX_2) {
            $this->buf->write($sign | WireConstants::BYTES_LEN_2, ($length >> 8) & 0xFF, $length & 0xFF, ...$data);
        } else {
            throw new \Exception('Bytes too long');
        }
        return $this->buf->length() - $start;
    }

    public function encodeArrayPayload(array $payload): int {
        return $this->encodeContainer($payload, Prefix::CONTAINER | WireConstants::CONTAINER_ARRAY);
    }

    public function encodeObjectPayload(array $payload): int {
        return $this->encodeContainer($payload, Prefix::CONTAINER | WireConstants::CONTAINER_MAP);
    }

    private function encodeContainer(array $payload, int $baseSign): int {
        $length = count($payload);
        $start = $this->buf->length();
        if ($length < WireConstants::CONTAINER_LEN_1) {
            $this->buf->write($baseSign | $length, ...$payload);
        } elseif ($length < WireConstants::MAX_1) {
            $this->buf->write($baseSign | WireConstants::CONTAINER_LEN_1, $length, ...$payload);
        } elseif ($length < WireConstants::MAX_2) {
            $this->buf->write($baseSign | WireConstants::CONTAINER_LEN_2, ($length >> 8) & 0xFF, $length & 0xFF, ...$payload);
        } else {
            throw new \Exception('Container payload too long');
        }
        return $this->buf->length() - $start;
    }

    public function encodeTagInner(array $tagBytes): int {
        if (empty($tagBytes)) {
            return 0;
        }
        if (count($tagBytes) > WireConstants::MAX_2) {
            throw new \Exception('Tag too long');
        }
        $start = $this->buf->length();
        $length = count($tagBytes);
        if ($length < 254) {
            $this->buf->write($length, ...$tagBytes);
        } elseif ($length < 257) {
            $this->buf->write(254, $length, ...$tagBytes);
        } else {
            $this->buf->write(255, ($length >> 8) & 0xFF, $length & 0xFF, ...$tagBytes);
        }
        return $this->buf->length() - $start;
    }

    public function encodeTaggedPayload(array $payload, array $rawTagFields): int {
        if (empty($rawTagFields)) {
            $this->buf->writeAll($payload);
            return count($payload);
        }
        $tEnc = new WireEncoder();
        $tEnc->encodeTagInner($rawTagFields);
        $tagEncoded = $tEnc->toByteArray();
        $length = count($tagEncoded) + count($payload);
        if ($length > WireConstants::MAX_2) {
            throw new \Exception('Tag+payload too long');
        }
        $start = $this->buf->length();
        $sign = Prefix::TAG;
        if ($length < WireConstants::TAG_LEN_1) {
            $this->buf->write($sign | $length, ...$tagEncoded, ...$payload);
        } elseif ($length < WireConstants::MAX_1) {
            $this->buf->write($sign | WireConstants::TAG_LEN_1, $length, ...$tagEncoded, ...$payload);
        } else {
            $this->buf->write($sign | WireConstants::TAG_LEN_2, ($length >> 8) & 0xFF, $length & 0xFF, ...$tagEncoded, ...$payload);
        }
        return $this->buf->length() - $start;
    }

    public function encodeBigIntDecimal(string $s): int {
        $bits = BigIntWireCodec::encodeSignedDecimal($s);
        $inner = array_merge([strlen($s)], $bits);
        return $this->encodeBytes($inner);
    }

    public function sliceFrom(int $start): int {
        return $this->buf->length() - $start;
    }

    public function copyLast(int $n): array {
        $end = $this->buf->length();
        return $this->buf->copyRange($end - $n, $end);
    }
}

File: mm-kt/src/main/kotlin/io/github/metamessage/core/WireEncoder.kt
package io.github.metamessage.core

import java.nio.charset.StandardCharsets

class WireEncoder {

    private val buf = GrowableByteBuf()

    fun toByteArray(): ByteArray = buf.copyRange(0, buf.length())

    fun finishTakeLast(writtenFromEnd: Int): ByteArray {
        val end = buf.length()
        return buf.copyRange(end - writtenFromEnd, end)
    }

    fun reset() {
        buf.reset()
    }

    fun size(): Int = buf.length()

    fun encodeSimple(simpleValue: Int): Int {
        require(simpleValue in 0..31) { "SimpleValue out of range: $simpleValue" }
        return writeByte((Prefix.SIMPLE or simpleValue).toByte())
    }

    fun encodeBool(v: Boolean): Int = encodeSimple(if (v) SimpleValue.TRUE else SimpleValue.FALSE)

    fun encodeInt64(v: Long): Int {
        if (v >= 0) {
            return encodeUintWithPrefix(Prefix.POSITIVE_INT, v)
        }
        val uv = if (v == Long.MIN_VALUE) 1L shl 63 else -v
        return encodeUintWithPrefix(Prefix.NEGATIVE_INT, uv)
    }

    fun encodeUint64(uv: Long): Int {
        require(uv >= 0) { "expected unsigned" }
        return encodeUintWithPrefix(Prefix.POSITIVE_INT, uv)
    }

    private fun encodeUintWithPrefix(prefix: Int, uv: Long): Int {
        val start = buf.length()
        when {
            java.lang.Long.compareUnsigned(uv, WireConstants.INT_LEN_1.toLong()) < 0 ->
                    buf.write((prefix or uv.toInt()).toByte())
            java.lang.Long.compareUnsigned(uv, WireConstants.MAX_1) <= 0 -> {
                buf.write((prefix or WireConstants.INT_LEN_1).toByte())
                buf.write(uv.toByte())
            }
            java.lang.Long.compareUnsigned(uv, WireConstants.MAX_2) <= 0 -> {
                buf.write((prefix or WireConstants.INT_LEN_2).toByte())
                buf.write((uv shr 8).toByte())
                buf.write(uv.toByte())
            }
            java.lang.Long.compareUnsigned(uv, WireConstants.MAX_3) <= 0 -> {
                buf.write((prefix or WireConstants.INT_LEN_3).toByte())
                buf.write((uv shr 16).toByte())
                buf.write((uv shr 8).toByte())
                buf.write(uv.toByte())
            }
            java.lang.Long.compareUnsigned(uv, WireConstants.MAX_4) <= 0 -> {
                buf.write((prefix or WireConstants.INT_LEN_4).toByte())
                buf.write((uv shr 24).toByte())
                buf.write((uv shr 16).toByte())
                buf.write((uv shr 8).toByte())
                buf.write(uv.toByte())
            }
            java.lang.Long.compareUnsigned(uv, WireConstants.MAX_5) <= 0 -> {
                buf.write((prefix or WireConstants.INT_LEN_5).toByte())
                buf.write((uv shr 32).toByte())
                buf.write((uv shr 24).toByte())
                buf.write((uv shr 16).toByte())
                buf.write((uv shr 8).toByte())
                buf.write(uv.toByte())
            }
            java.lang.Long.compareUnsigned(uv, WireConstants.MAX_6) <= 0 -> {
                buf.write((prefix or WireConstants.INT_LEN_6).toByte())
                buf.write((uv shr 40).toByte())
                buf.write((uv shr 32).toByte())
                buf.write((uv shr 24).toByte())
                buf.write((uv shr 16).toByte())
                buf.write((uv shr 8).toByte())
                buf.write(uv.toByte())
            }
            java.lang.Long.compareUnsigned(uv, WireConstants.MAX_7) <= 0 -> {
                buf.write((prefix or WireConstants.INT_LEN_7).toByte())
                buf.write((uv shr 48).toByte())
                buf.write((uv shr 40).toByte())
                buf.write((uv shr 32).toByte())
                buf.write((uv shr 24).toByte())
                buf.write((uv shr 16).toByte())
                buf.write((uv shr 8).toByte())
                buf.write(uv.toByte())
            }
            else -> {
                buf.write((prefix or WireConstants.INT_LEN_8).toByte())
                buf.write((uv shr 56).toByte())
                buf.write((uv shr 48).toByte())
                buf.write((uv shr 40).toByte())
                buf.write((uv shr 32).toByte())
                buf.write((uv shr 24).toByte())
                buf.write((uv shr 16).toByte())
                buf.write((uv shr 8).toByte())
                buf.write(uv.toByte())
            }
        }
        return buf.length() - start
    }

    fun encodeFloatString(s: String): Int {
        val p = FloatCodec.parseDecimalString(s)
        val start = buf.length()
        var sign = Prefix.FLOAT
        if (p.negative) sign = sign or WireConstants.FLOAT_NEG_MASK
        if (p.exponent.toInt() == -1 && p.mantissa <= 7) {
            writeByte((sign or p.mantissa.toInt()).toByte())
        } else {
            val mantissa = p.mantissa
            when {
                mantissa <= WireConstants.MAX_1 -> {
                    writeByte((sign or WireConstants.FLOAT_LEN_1).toByte())
                    writeByte(p.exponent)
                    writeByte(mantissa.toByte())
                }
                mantissa <= WireConstants.MAX_2 -> {
                    writeByte((sign or WireConstants.FLOAT_LEN_2).toByte())
                    writeByte(p.exponent)
                    writeByte((mantissa shr 8).toByte())
                    writeByte(mantissa.toByte())
                }
                mantissa <= WireConstants.MAX_3 -> {
                    writeByte((sign or WireConstants.FLOAT_LEN_3).toByte())
                    writeByte(p.exponent)
                    writeByte((mantissa shr 16).toByte())
                    writeByte((mantissa shr 8).toByte())
                    writeByte(mantissa.toByte())
                }
                mantissa <= WireConstants.MAX_4 -> {
                    writeByte((sign or WireConstants.FLOAT_LEN_4).toByte())
                    writeByte(p.exponent)
                    writeByte((mantissa shr 24).toByte())
                    writeByte((mantissa shr 16).toByte())
                    writeByte((mantissa shr 8).toByte())
                    writeByte(mantissa.toByte())
                }
                mantissa <= WireConstants.MAX_5 -> {
                    writeByte((sign or WireConstants.FLOAT_LEN_5).toByte())
                    writeByte(p.exponent)
                    writeByte((mantissa shr 32).toByte())
                    writeByte((mantissa shr 24).toByte())
                    writeByte((mantissa shr 16).toByte())
                    writeByte((mantissa shr 8).toByte())
                    writeByte(mantissa.toByte())
                }
                mantissa <= WireConstants.MAX_6 -> {
                    writeByte((sign or WireConstants.FLOAT_LEN_6).toByte())
                    writeByte(p.exponent)
                    writeByte((mantissa shr 40).toByte())
                    writeByte((mantissa shr 32).toByte())
                    writeByte((mantissa shr 24).toByte())
                    writeByte((mantissa shr 16).toByte())
                    writeByte((mantissa shr 8).toByte())
                    writeByte(mantissa.toByte())
                }
                mantissa <= WireConstants.MAX_7 -> {
                    writeByte((sign or WireConstants.FLOAT_LEN_7).toByte())
                    writeByte(p.exponent)
                    writeByte((mantissa shr 48).toByte())
                    writeByte((mantissa shr 40).toByte())
                    writeByte((mantissa shr 32).toByte())
                    writeByte((mantissa shr 24).toByte())
                    writeByte((mantissa shr 16).toByte())
                    writeByte((mantissa shr 8).toByte())
                    writeByte(mantissa.toByte())
                }
                else -> {
                    writeByte((sign or WireConstants.FLOAT_LEN_8).toByte())
                    writeByte(p.exponent)
                    writeByte((mantissa shr 56).toByte())
                    writeByte((mantissa shr 48).toByte())
                    writeByte((mantissa shr 40).toByte())
                    writeByte((mantissa shr 32).toByte())
                    writeByte((mantissa shr 24).toByte())
                    writeByte((mantissa shr 16).toByte())
                    writeByte((mantissa shr 8).toByte())
                    writeByte(mantissa.toByte())
                }
            }
        }
        return buf.length() - start
    }

    fun encodeString(s: String): Int {
        val utf = s.toByteArray(StandardCharsets.UTF_8)
        val length = utf.size
        require(length <= WireConstants.MAX_2) { "string too long" }
        val sign = Prefix.STRING
        return when {
            length < WireConstants.STRING_LEN_1 ->
                    writeBytesWithPrefix(utf, (sign or length).toByte())
            length < WireConstants.MAX_1 ->
                    writeBytesWithPrefix(
                            utf,
                            (sign or WireConstants.STRING_LEN_1).toByte(),
                            length.toByte()
                    )
            else ->
                    writeBytesWithPrefix(
                            utf,
                            (sign or WireConstants.STRING_LEN_2).toByte(),
                            (length shr 8).toByte(),
                            length.toByte()
                    )
        }
    }

    fun encodeBytes(data: ByteArray): Int {
        val length = data.size
        require(length <= WireConstants.MAX_2) { "bytes too long" }
        val sign = Prefix.BYTES
        return when {
            length < WireConstants.BYTES_LEN_1 ->
                    writeBytesWithPrefix(data, (sign or length).toByte())
            length < WireConstants.MAX_1 ->
                    writeBytesWithPrefix(
                            data,
                            (sign or WireConstants.BYTES_LEN_1).toByte(),
                            length.toByte()
                    )
            else ->
                    writeBytesWithPrefix(
                            data,
                            (sign or WireConstants.BYTES_LEN_2).toByte(),
                            (length shr 8).toByte(),
                            length.toByte()
                    )
        }
    }

    fun encodeArrayPayload(payload: ByteArray): Int =
            encodeContainer(payload, Prefix.CONTAINER or WireConstants.CONTAINER_ARRAY)

    fun encodeObjectPayload(payload: ByteArray): Int =
            encodeContainer(payload, Prefix.CONTAINER or WireConstants.CONTAINER_MAP)

    private fun encodeContainer(payload: ByteArray, baseSign: Int): Int {
        val length = payload.size
        require(length <= WireConstants.MAX_2) { "container payload too long" }
        return when {
            length < WireConstants.CONTAINER_LEN_1 ->
                    writeBytesWithPrefix(payload, (baseSign or length).toByte())
            length < WireConstants.MAX_1 ->
                    writeBytesWithPrefix(
                            payload,
                            (baseSign or WireConstants.CONTAINER_LEN_1).toByte(),
                            length.toByte()
                    )
            else ->
                    writeBytesWithPrefix(
                            payload,
                            (baseSign or WireConstants.CONTAINER_LEN_2).toByte(),
                            (length shr 8).toByte(),
                            length.toByte()
                    )
        }
    }

    fun encodeTagInner(tagBytes: ByteArray): Int {
        if (tagBytes.isEmpty()) return 0
        require(tagBytes.size <= WireConstants.MAX_2) { "tag too long" }
        val length = tagBytes.size
        return when {
            length < 254 -> writeBytesWithPrefix(tagBytes, length.toByte())
            length < 257 -> writeBytesWithPrefix(tagBytes, 254.toByte(), length.toByte())
            else ->
                    writeBytesWithPrefix(
                            tagBytes,
                            255.toByte(),
                            (length shr 8).toByte(),
                            length.toByte()
                    )
        }
    }

    fun encodeTaggedPayload(payload: ByteArray, rawTagFields: ByteArray): Int {
        if (rawTagFields.isEmpty()) {
            writeBytes(payload)
            return payload.size
        }
        val tEnc = WireEncoder()
        tEnc.encodeTagInner(rawTagFields)
        val tagEncoded = tEnc.toByteArray()
        val length = tagEncoded.size + payload.size
        require(length <= WireConstants.MAX_2) { "tag+payload too long" }
        val start = buf.length()
        val sign = Prefix.TAG
        when {
            length < WireConstants.TAG_LEN_1 -> {
                buf.write((sign or length).toByte())
                buf.writeAll(tagEncoded)
                buf.writeAll(payload)
            }
            length < WireConstants.MAX_1 -> {
                buf.write((sign or WireConstants.TAG_LEN_1).toByte())
                buf.write(length.toByte())
                buf.writeAll(tagEncoded)
                buf.writeAll(payload)
            }
            else -> {
                buf.write((sign or WireConstants.TAG_LEN_2).toByte())
                buf.write((length shr 8).toByte())
                buf.write(length.toByte())
                buf.writeAll(tagEncoded)
                buf.writeAll(payload)
            }
        }
        return buf.length() - start
    }

    fun encodeBigIntDecimal(s: String): Int {
        val bits = BigIntWireCodec.encodeSignedDecimal(s)
        val inner = ByteArray(1 + bits.size)
        inner[0] = s.length.toByte()
        System.arraycopy(bits, 0, inner, 1, bits.size)
        return encodeBytes(inner)
    }

    fun sliceFrom(start: Int): Int = buf.length() - start

    fun copyLast(n: Int): ByteArray {
        val end = buf.length()
        return buf.copyRange(end - n, end)
    }

    private fun writeByte(vararg bs: Byte): Int {
        buf.write(*bs)
        return bs.size
    }

    private fun writeBytes(bs: ByteArray): Int {
        buf.writeAll(bs)
        return bs.size
    }

    private fun writeBytesWithPrefix(bs: ByteArray, vararg prefix: Byte): Int {
        for (b in prefix) {
            buf.write(b)
        }
        buf.writeAll(bs)
        return prefix.size + bs.size
    }
}
